modules/ssh/init.zsh

```diff
@@ -14,13 +14,32 @@ fi
 _ssh_dir="$HOME/.ssh"
 
 # Set the path to the environment file if not set by another module.
-_ssh_agent_env="${_ssh_agent_env:-${TMPDIR:-/tmp}/ssh-agent.env}"
+_ssh_agent_env="${_ssh_agent_env:-${TMPDIR:-/tmp}/ssh-agent-"$LOGNAME".env}"
 
-# Start ssh-agent if not started.
+# Due to the predictability of the env file, check the env file exists and is
+# owned by current EUID before trusting it.
+if [ -f "$_ssh_agent_env" -a ! -O "$_ssh_agent_env" ]; then
+  cat 1>&2 <<-EOF
+	ERROR: Cannot trust the SSH agent environment variables persistence
+	file because it is owned by another user.
+	The ssh-agent will not be started.
+	$_ssh_agent_env
+	EOF
+  unset _ssh_{dir,agent_env}
+  return 1
+fi
+
+# If a socket exists at SSH_AUTH_SOCK, assume ssh-agent is already running and
+# skip starting it.
 if [[ ! -S "$SSH_AUTH_SOCK" ]]; then
-  # Export environment variables.
+  # Try to grab previously exported environment variables.
   source "$_ssh_agent_env" 2> /dev/null
-  eval "$(ssh-agent -s)"
+
+  # Do not start ssh-agent if the PID from the last start of ssh-agent exists and
+  # corresponds to a running ssh-agent under the current user.
+  if ! ps -U "$LOGNAME" -o pid,ucomm | grep -q -- "${SSH_AGENT_PID:--1} ssh-agent"; then
```

Comment:
I took the ps+grep pipeline from a previous version of this ssh module.  The criteria being tested is that an `ssh-agent` process for the current user is running with the expected PID.

Using `pgrep` would require some different logic since there is no command line arg to pass a PID for matching.  One possibility would be `[ "$(pgrep -U $LOGNAME -n -x ssh-agent)" -eq "${SSH_AGENT_PID:--1}" ]`

To my knowledge, [ps](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/ps.html)|[grep](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) is more portable as its POSIX.  I cannot find pgrep in POSIX.

One thing to note is that, after checking the POSIX spec, `ucomm` should be changed to `comm`.